code/Serving_API.py

```diff
@@ -1,27 +1,258 @@
+            G, source=src_city_info, target=dst_city_info, weight='distance')
+
+        shortest_distance, coordinate_list, wkt_list, cable_type_list = calculate_shortest_path_distance(
+            G, shortest_path_cities)
+
+        return (coordinate_list, wkt_list, cable_type_list)
+    except nx.NetworkXNoPath:
+        print("cannot find the shortest path")
+        return ([], [], [])  # or handle the error as you prefer
+
 
 def run():
     import uvicorn
     uvicorn.run(app, port=8082)
 
+
+def calculate_shortest_path_distance(G, shortest_path_cities):
+    total_distance = 0
+    coordinate_list = []
+    wkt_list = []
+    cable_type_list = []
+
+    for i in range(len(shortest_path_cities) - 1):
+        city1 = shortest_path_cities[i]
+        city2 = shortest_path_cities[i + 1]
+
+        total_distance += G[city1][city2]['weight']
+        # Only append the coordinates of the source city here
+        coordinate_list.append(G.nodes[city1]['coord'])
+        wkt_list.append(G[city1][city2]['path_wkt'])
+        cable_type_list.append(G[city1][city2]['cable_type'])
+
+    # Append the coordinates of the last city after the loop
+    coordinate_list.append(G.nodes[shortest_path_cities[-1]]['coord'])
+
+    return total_distance, coordinate_list, wkt_list, cable_type_list
+
+
+def city_formatter(city_info):
+    city, state, country = city_info
+    return (city.strip(), state.strip(), country.strip())
+
+
+def parse_csv(filename):
+    data = []
+    with open(filename, newline='') as csvfile:
+        reader = csv.DictReader(csvfile)
+        for row in reader:
+            data.append(row)
+    return data
+
+
+def function_tester():
+    filename = 'all_pairs.by_geo.csv'
+    parsed_data = parse_csv(filename)
+    for item in parsed_data:
+        src_city_co = (float(item['src_latitude']),
+                       float(item['src_longitude']))
+        dst_city_co = (float(item['dst_latitude']),
+                       float(item['dst_longitude']))
+
+        print(src_city_co)
+        print(dst_city_co)
+
+        src_city_co = find_closest_point(
+            src_city_co, coord_set)
+
+        dst_city_co = find_closest_point(
+            dst_city_co, coord_set)
+
+        src_city_info = coordCityMap[src_city_co]
+
+        dst_city_info = coordCityMap[dst_city_co]
+
+        shortest_path_cities = nx.shortest_path(
+            G, source=src_city_info, target=dst_city_info, weight='distance')
+
+        shortest_distance, coordinate_list, wkt_list, cable_type_list = calculate_shortest_path_distance(
+            G, shortest_path_cities)
+
+        print(src_city_co)
+        print(dst_city_co)
+
+        print(src_city_info)
+        print(dst_city_info)
+
+        print(wkt_list)
+        print(cable_type_list)
+        print(coordinate_list)
+        print("\n")
+
+
 if __name__ == "__main__":
+    db_file = '../database/igdb.db'
+    coordCityMap, coord_set, G = build_up_global_graph(db_file)
```

Comment:
Move the initialization code inside `run()` and assign global variables to app as an attribute, like [this](https://github.com/tiangolo/fastapi/issues/592#issuecomment-539419645)

Reply:
fixed